dss/api/collections.py

```diff
@@ -0,0 +1,190 @@
+        raise DSSException(requests.codes.forbidden, "forbidden", f"Collection access denied")
+
+    blobstore = Config.get_blobstore_handle(Replica[replica])
+    bucket = Replica[replica].bucket
+    created, idempotent = _idempotent_save(blobstore, bucket, tombstone_key, tombstone_object_data)
+    if not idempotent:
+        raise DSSException(requests.codes.conflict,
+                           f"collection_tombstone_already_exists",
+                           f"collection tombstone with UUID {uuid} already exists")
+    status_code = requests.codes.ok
+    response_body = dict()  # type: dict
+
+    return jsonify(response_body), status_code
+
+@functools.lru_cache(maxsize=64)
+def get_json_metadata(entity_type: str, uuid: str, version: str, replica: Replica, blobstore_handle: BlobStore):
+    try:
+        key = "{}s/{}.{}".format(entity_type, uuid, version)
+        # TODO: verify that file is a metadata file
+        size = blobstore_handle.get_size(replica.bucket, key)
+        if size > MAX_METADATA_SIZE:
+            raise DSSException(422, "invalid_link",
+                               "The file UUID {} refers to a file that is too large to process".format(uuid))
+        return json.loads(blobstore_handle.get(
+            replica.bucket,
+            "{}s/{}.{}".format(entity_type, uuid, version)))
+    except BlobNotFoundError as ex:
+        raise DSSException(404, "invalid_link", "Could not find file for UUID {}".format(uuid))
+
+def resolve_content_item(replica: Replica, blobstore_handle: BlobStore, item: dict):
+    try:
+        if item["type"] in {"file", "bundle", "collection"}:
+            item_metadata = get_json_metadata(item["type"], item["uuid"], item["version"], replica, blobstore_handle)
+        else:
+            item_metadata = get_json_metadata("file", item["uuid"], item["version"], replica, blobstore_handle)
+            if "fragment" not in item:
+                raise Exception('The "fragment" field is required in collection elements '
+                                'other than files, bundles, and collections')
+            blob_path = "blobs/" + ".".join((
+                item_metadata[FileMetadata.SHA256],
+                item_metadata[FileMetadata.SHA1],
+                item_metadata[FileMetadata.S3_ETAG],
+                item_metadata[FileMetadata.CRC32C],
+            ))
+            # check that item is marked as metadata, is json, and is less than max size
+            item_doc = json.loads(blobstore_handle.get(replica.bucket, blob_path))
+            item_content = jsonpointer.resolve_pointer(item_doc, item["fragment"])
+            return item_content
+    except DSSException:
+        raise
+    except Exception as e:
+        raise DSSException(
+            422,
+            "invalid_link",
+            'Error while parsing the link "{}": {}: {}'.format(item, type(e).__name__, e)
+        )
+
+def verify_collection(contents: List[dict], replica: Replica, blobstore_handle: BlobStore):
```

Comment:
Suggestion: It seems that PATCH ultimately assumes that a collection is a set of items. I think we should therefore ensure uniqueness of collection items in PUT as well.

Reply:
I agree, implementing that.

Reply:
Done.